### ISSUE TITLE
Feature: Theming

### ISSUE THREAD

**Issue by rdbende** [tukaan/tukaan#62]
TODO:

- [ ] Apply native theme
      Would be nice to use [tile-gtk](https://github.com/rdbende/gttk) (and tile-qt) on Linux, though I can't release binaries here, as Gtk is LGPL
- [ ] Get current theme
- [ ] Set theme
- [ ] Get available themes

**Comment by Moosems** [comment]
Can we add the sun-valley ttk theme as a built-in theme? Question: can we still allow the user to choose button width/height and font even when using a special theme (something tkinter does not allow)?

**Comment by sumeshir26** [comment]
> Can we add the sun-valley ttk theme as a built-in theme? Question: can we still allow the user to choose button width/height and font even when using a special theme (something tkinter does not allow)?

@rdbende said he was going to make a extension for Tukkan with many themes

**Comment by sumeshir26** [comment]
So, if I have a Windows 11 GTK/KDE theme, will the app match that theme, or does GTTK only work with system themes?

Sorry this is a stupid question, but I don't have Linux so I was curious.

**Comment by rdbende** [comment]
Gttk can use any Gtk theme that is installed on the system, provided that the theme has a Gtk 2.0 version. So if it's Gtk 3 only (which uses CSS for styling), it won't be able to use it.
On KDE Plasma (which is based on Qt) it can't use the widget styles, instead it will use the buil-in clam theme, and apply the system colorscheme to it

**Comment by Moosems** [comment]
When we add theming can we still allow the user to choose button width/height and font even when using a special theme (something tkinter does not allow)? I find it important to be able to choose font because I brand my apps with a special font style but also love ttk looks so it's often really frustrating for me to not be able to get both (A silly annoyance I know)

**Comment by sumeshir26** [comment]
> Gttk can use any Gtk theme that is installed on the system, provided that the theme has a Gtk 2.0 version. So if it's Gtk 3 only (which uses CSS for styling), it won't be able to use it. On KDE Plasma (which is based on Qt) it can't use the widget styles, instead it will use the buil-in clam theme, and apply the system colorscheme to it

I thought Tukaan also supports CSS theming?

**Comment by rdbende** [comment]
Yes, but the Tukaan css theming is a bit different than Gtk.
In Tikaan you have to use png images, just as in Tkinter, but the css syntax makes it much easier to set the other properties, however it doesn't support dynamic styling.

**Comment by Moosems** [comment]
What do you mean by dynamic styling?

**Comment by rdbende** [comment]
When you do this:
```css
.button {
  color: #fafafa;
  background-color: #6667ab;
  padding: 4px 8px;
  border-radius: 4px;
}
```

Tukaan can't generate widget styles based on information like this.

**Comment by Moosems** [comment]
I could create a lexer for it. We could also have it use a png generator that bases it off of a built in design and modify it to specifications. It would be a bit of work but could be done

**Comment by rdbende** [comment]
Yupp, that was my first idea when I started with ttk themes. But then it became "just" the Azure theme, because I realized how much work it would be just to build a CSS parser and generate Tcl scripts from it. However, if I'm actually halfway done with that whole thing in Tukaan, why not make a png generator too? (I believe, PIL.ImageDraw doesn't support antialiasing, so it wouldn't be suitable for this purpose.)

**Comment by rdbende** [comment]
They **do** support antialiasing on every platform! 🤔
Just take a look at one of my themes, they are all antialiased.
Tk supports displaying antialiased images, and since most ttk theme are mafe of pixmap, they look good even on Windows. The only thing in Tk that doesn't support antialiasing on Linux and Windows is the `Canvas` widget, but (of course :) Tukaan has a alternative canvas implementation, that has everything you can imagine.

**Comment by alejandroautalan** [comment]
Hello,

> Tukaan has a alternative canvas implementation, that has everything you can imagine.

I am interested in the implementation of the canvas and its improvements. It's not released yet right?

**Comment by rdbende** [comment]
Yup, it's not yet complete. But basically it just uses the [TkPath](https://github.com/avarga/tkpath) project (an alternative canvas widget that uses native backends for drawing), instead of Tk's original canvas widget. So actually it's not the original canvas that's been improved, but it's an alternative implementation.
I'm also working on a wrapper around the `TkPath` canvas widget for Tkinter.

**Comment by Moosems** [comment]
What does the canvas look like currently?

**Comment by rdbende** [comment]
```python
import tukaan
from tukaan.gfx import Brush, Canvas, Gradient, Pen, Transform


app = tukaan.App()

c = Canvas(app)
c.grid()

rect = Canvas.Rect(
    c,
    30,
    30,
    80,
    80,
    border_radius=10,
    brush=Brush(fill=Gradient.Linear(["#cc00ff", "#007fff"])),
    pen=Pen(color="#fff", opacity=0.7, width=2),
)

path = Canvas.Path(
    c,
    draw="M 50 50 C 50 100, 50 100, 100 50",
    pen=Pen(color="#aaffaa", width=4, line_cap="round"),
)

rect.transform(Transform.Scale(1.3) @ Transform.TranslateX(-10))


app.run()
```

![image](https://user-images.githubusercontent.com/77941087/177360644-07dbbb81-d432-4e37-96d4-01f667fdf18a.png)

**Comment by Moosems** [comment]
Can you break down what path is?

**Comment by rdbende** [comment]
Path is an SVG element originally. Basically it's a very powerful pen element with which you can draw lines, curves, arcs, and more. 

Documentation: https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Paths

**Comment by Moosems** [comment]
I have a feeling that this week need a very in depth segment of documentation

**Comment by rdbende** [comment]
https://tukaan.github.io/docs/themes/themes

**Comment by sumeshir26** [comment]
> https://tukaan.github.io/docs/themes/themes

Theres nothing over there...
![image](https://user-images.githubusercontent.com/68823982/209346571-366b5af1-ba13-4a9c-bb28-45f6c6ca385d.png)

**Comment by rdbende** [comment]
Yes, I know. That's why we need to fill it in 🙃

**Comment by rdbende** [comment]
But wait, is that little thingy under the "Docs" label actually dark blue?

**Comment by sumeshir26** [comment]
> But wait, is that little thingy under the "Docs" label actually dark blue?

Yes

**Comment by rdbende** [comment]
That's strange. It should be light blue, like the edit button on the tight. It's that light blue color for me on both desktop and mobile and I even tried it in another browser (Chromium).

**Comment by rdbende** [comment]
> No - its light blue...
(At least if i check now)

👍
